File: Make_Array_Strictly_Increasing.py
class Solution:
    def makeArrayIncreasing(self, arr1: List[int], arr2: List[int]) -> int:
        prev = {-1: 0}
        arr2.sort()

        def binary_search(target):
            left = 0
            right = len(arr2) - 1
            
            while left <= right:
                mid = (left + right) // 2
                
                if arr2[mid] <= target:
                    left = mid + 1
                else:
                    right = mid - 1
            
            return left

        for n in arr1:
            temp = defaultdict(lambda: float('inf'))

            for key in prev:
                if n > key:
                    temp[n] = min(temp[n], prev[key])
                
                idx = binary_search(key)

                if idx < len(arr2):
                    temp[arr2[idx]] = min(temp[arr2[idx]], prev[key] + 1)

            prev = temp

        return min(prev.values()) if prev else -1

File: Make_Array_Strictly_Increasing.java
class Solution {
    public int makeArrayIncreasing(int[] arr1, int[] arr2) {
        Map<Integer, Integer> prev = new HashMap<>();
        prev.put(-1, 0);
        Arrays.sort(arr2);

        for (int n : arr1) {
            Map<Integer, Integer> temp = new HashMap<>();

            for (int key : prev.keySet()) {
                if (n > key) {
                    temp.put(n, Math.min(temp.getOrDefault(n, Integer.MAX_VALUE), prev.get(key)));
                }
                
                int idx = binarySearch(key, arr2);
                
                if (idx < arr2.length) {
                    temp.put(arr2[idx], Math.min(temp.getOrDefault(arr2[idx], Integer.MAX_VALUE), prev.get(key) + 1));
                }
            }

            prev = temp;
        }

        return prev.values().stream().min(Integer::compareTo).orElse(-1);
    }
    
    private int binarySearch(int target, int[] arr) {
        int left = 0;
        int right = arr.length - 1;
        
        while (left <= right) {
            int mid = (left + right) / 2;
            
            if (arr[mid] <= target) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        
        return left;
    }        
}
